transport/dnsovertcp/packet_listener.go

```diff
@@ -0,0 +1,349 @@
+	dnsTcpSizeBufferLen = 2           // Buffer length (2 bytes) of the prefixed size in TCP
+	dnsUdpMinMsgLen     = 12          // A DNS message must at least containing the header
+	dnsUdpMaxMsgLen     = 512         // https://datatracker.ietf.org/doc/html/rfc1035#section-2.3.4
+	dnsUdpTruncatedByte = 2           // The byte in the header containing TC bit
+	dnsUdpTruncatedBit  = uint8(0x02) // The TC bit within dnsTruncatedByte
+)
+
+type dnsOverTcpHandler struct {
+	sd transport.StreamDialer
+}
+
+type dnsResult struct {
+	p    []byte
+	n    int
+	addr net.Addr
+	err  error
+	wg   *sync.WaitGroup
+}
+
+type dnsOverTcpConn struct {
+	mu   sync.Mutex
+	done chan struct{}
+
+	sd    transport.StreamDialer
+	conns *list.List
+	wrDdl atomic.Pointer[time.Time]
+
+	dnsRecv chan *dnsResult
+	recvDdl *ddltimer.DeadlineTimer
+}
+
+// Compilation guard against interface implementation
+var _ transport.PacketListener = (*dnsOverTcpHandler)(nil)
+var _ net.PacketConn = (*dnsOverTcpConn)(nil)
+
+func NewPacketListener(sd transport.StreamDialer) (transport.PacketListener, error) {
+	return &dnsOverTcpHandler{
+		sd: sd,
+	}, nil
+}
+
+func (h *dnsOverTcpHandler) ListenPacket(ctx context.Context) (net.PacketConn, error) {
+	conn := &dnsOverTcpConn{
+		done:    make(chan struct{}),
+		sd:      h.sd,
+		conns:   list.New(),
+		dnsRecv: make(chan *dnsResult),
+		recvDdl: ddltimer.New(),
+	}
+	conn.wrDdl.Store(&time.Time{})
+	return conn, nil
+}
+
+func (c *dnsOverTcpConn) Close() (err error) {
+	// Lock() for safely reading the `c.conns` list
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	err = c.forEachConnIgnoreErrClosed(func(conn net.Conn) error {
+		return conn.Close()
+	})
+
+	c.recvDdl.Stop()
+
+	select {
+	case <-c.done:
+		// prevent panic of closing a closed channel
+	default:
+		close(c.done)
+	}
+	return
+}
+
+func (c *dnsOverTcpConn) ReadFrom(p []byte) (n int, addr net.Addr, err error) {
+	if c.alreadyClosed() {
+		return 0, nil, io.EOF
+	}
+	if len(p) == 0 {
+		return 0, nil, errShortBuffer
+	}
+
+	recv := &dnsResult{
+		p:  p,
+		wg: &sync.WaitGroup{},
+	}
+	recv.wg.Add(1)
+
+	// Wait until either one of the following:
+	//   1. one of the `listenDNSResponseOverTCP` goroutine responded through recv.wg.Done()
+	//   2. c is closed
+	//   3. read timed out (`listenDNSResponseOverTCP` won't respond if conn.Read() timed out)
+	select {
+	case c.dnsRecv <- recv:
+		recv.wg.Wait()
+	case <-c.done:
+		return 0, nil, io.EOF
+	case <-c.recvDdl.Timeout():
+		return 0, nil, errTimedOut
+	}
+
+	n, addr, err = recv.n, recv.addr, recv.err
+	if errors.Is(err, net.ErrClosed) {
+		err = io.EOF
+	}
+	return
+}
+
+func (c *dnsOverTcpConn) WriteTo(p []byte, addr net.Addr) (n int, err error) {
+	if c.alreadyClosed() {
+		return 0, errClosed
+	}
+	if udpAddr, ok := addr.(*net.UDPAddr); !ok || udpAddr.Port != dnsServerPort {
+		return 0, errInvalidResolver
+	}
+	if len(p) < dnsUdpMinMsgLen {
+		return 0, errInvalidDNSMsg
+	}
+
+	conn, n, err := c.sendDNSRequestOverTCP(p, addr)
```

Comment:
This waits on a TCP connection and will cause head-of-line blocking. You won't be able to process the next UDP packet until this write returns, while you could be doing that in parallel.

Move the Dial+Write to the Go routine. You already have the `buf` in `listenDNSResponseOverTCP`. Copy `p` there before starting the go-routine, and use it for the Write to the proxy.